org-code-javaide-codebuilder/lib/src/main/java/org/code/javaide/codebuilder/CodeExecutor.java

```diff
@@ -0,0 +1,97 @@
+package org.code.javaide.codebuilder;
+
+import com.amazonaws.services.lambda.runtime.Context;
+import com.amazonaws.services.lambda.runtime.RequestHandler;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.file.Paths;
+import java.util.Map;
+
+import org.apache.commons.io.FilenameUtils;
+
+/**
+ * This is the AWS Lambda handler
+ * Details: https://docs.aws.amazon.com/lambda/latest/dg/java-handler.html
+ * It compiles and runs the input Java code and returns the output of that code's execution
+ */
+public class CodeExecutor implements RequestHandler<Map<String,String>, String>{
+  Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+  /**
+  * @param userProject A mapping of filenames in a user's project to the code in those files
+  * @param __          Unused. The environment context where the handler is executed
+  * @return            The output from the user's code execution
+  */
+  @Override
+  public String handleRequest(Map<String,String> userProject, Context __)
+  {
+    // TODO: CSA-48 Handle more than one file
+    if(userProject.size() > 1) {
+      return "Error: Can only process one file at a time.";
+    }
+
+    File compileRunScript = null;
+    String tempDir = System.getProperty("java.io.tmpdir");
+    for(Map.Entry<String, String> projectFile : userProject.entrySet()) {
+      String className = FilenameUtils.removeExtension(projectFile.getKey());
+      String userCode = projectFile.getValue();
+      try {
+        // Build user's java file
+        File userFile = new File(Paths.get(tempDir, projectFile.getKey()).toString());
+        FileWriter writer = new FileWriter(userFile);
+        writer.write(userCode);
+        writer.close();
+
+        // Create the compile & run script
+        compileRunScript = File.createTempFile("script", null);
+        FileWriter scriptWriter = new FileWriter(compileRunScript);
+        scriptWriter.write("javac " + userFile.getAbsolutePath());
+        scriptWriter.write(System.getProperty( "line.separator" ));
+        scriptWriter.write("java -cp " + tempDir + " " + className);
+        scriptWriter.close();
+      } catch (IOException e) {
+        e.printStackTrace();
+        // TODO: CSA-42 Improve error handling
+        return "An error occurred creating files.";
+      }
+    }
+
+    ProcessBuilder pb = new ProcessBuilder().command("bash", compileRunScript.toString());
```

Comment:
I believe the Lambda execution environment is designed for the main “work” or functionality of the Lambda to be carried out by the RequestHandler and any code it invokes within that process. While other processes can be started within the Lambda container, the Lambda execution environment may not be able to ensure those other processes are cleaned up at the end of the RequestHandler’s execution. If those other processes hang or are still executing when the RequestHandler completes, they may persist and resume executing the next time the Lambda is invoked and the same container is re-used, causing unintended side effects. From the [docs](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-context.html):

```
Background processes or callbacks that were initiated by your Lambda function and did not complete when the function ended resume if Lambda reuses the execution environment. Make sure that any background processes or callbacks in your code are complete before the code exits.
```

It may be worth investigating whether we can use the [Java compiler Module](https://docs.oracle.com/en/java/javase/11/docs/api/java.compiler/module-summary.html) to compile the student’s source code, at which point we could possibly load/execute the student’s compiled classes within the main Lambda RequestHandler execution environment. Sand boxing the student’s Java code would be a challenge to ensure it doesn’t invoke/manipulate our own RequestHandler’s functionality.

Found some outdated examples of using `java.compiler`, but those might be useful in prototyping:

https://javabeat.net/the-java-6-0-compiler-api/
https://www.javacodegeeks.com/2015/09/java-compiler-api.html